wcfsetup/install/files/lib/data/file/File.class.php

```diff
@@ -0,0 +1,192 @@
+
+    /** @var array<string, FileThumbnail> */
+    private array $thumbnails = [];
+
+    public function getSourceFilename(): string
+    {
+        return \sprintf(
+            '%d-%s-%s.%s',
+            $this->fileID,
+            $this->secret,
+            $this->fileHash,
+            $this->fileExtension,
+        );
+    }
+
+    public function getPath(): string
+    {
+        $folderA = \substr($this->fileHash, 0, 2);
+        $folderB = \substr($this->fileHash, 2, 2);
+
+        return \sprintf(
+            \WCF_DIR . '_data/%s/files/%s/%s/',
+            $this->fileExtension === 'bin' ? 'private' : 'public',
+            $folderA,
+            $folderB,
+        );
+    }
+
+    public function getPathname(): string
+    {
+        return $this->getPath() . $this->getSourceFilename();
+    }
+
+    public function getLink(): string
+    {
+        return LinkHandler::getInstance()->getControllerLink(
+            FileDownloadAction::class,
+            ['id' => $this->fileID]
+        );
+    }
+
+    public function getProcessor(): ?IFileProcessor
+    {
+        return FileProcessor::getInstance()->getProcessorById($this->objectTypeID);
+    }
+
+    public function isImage(): bool
+    {
+        return match ($this->mimeType) {
+            'image/gif' => true,
+            'image/jpeg' => true,
+            'image/png' => true,
+            'image/webp' => true,
+            default => false,
+        };
+    }
+
+    public function canDelete(): bool
+    {
+        $processor = $this->getProcessor();
+        if ($processor === null) {
+            return true;
+        }
+
+        return $processor->canDelete($this);
+    }
+
+    public function addThumbnail(FileThumbnail $thumbnail): void
+    {
+        $this->thumbnails[$thumbnail->identifier] = $thumbnail;
+    }
+
+    public function getThumbnail(string $identifier): ?FileThumbnail
+    {
+        return $this->thumbnails[$identifier] ?? null;
+    }
+
+    public function toHtmlElement(array $metaData): string
+    {
+        $thumbnails = [];
+        foreach ($this->thumbnails as $thumbnail) {
+            $thumbnails[] = [
+                'identifier' => $thumbnail->identifier,
+                'link' => $thumbnail->getLink(),
+            ];
+        }
+
+        return \sprintf(
+            <<<'EOT'
+                <woltlab-core-file
+                    file-id="%d"
+                    data-filename="%s"
+                    data-file-size="%s"
+                    data-mime-type="%s"
+                    data-thumbnails="%s"
+                    data-meta-data="%s"
+                    data-link="%s"
+                ></woltlab-core-file>
+                EOT,
+            $this->fileID,
+            StringUtil::encodeHTML($this->filename),
+            $this->fileSize,
+            StringUtil::encodeHTML($this->mimeType),
+            StringUtil::encodeHTML(\json_encode($thumbnails)),
+            StringUtil::encodeHTML(\json_encode($metaData)),
```

Comment:
This should include the `JSON_THROW_ON_ERROR` flag if `JSON::encode()` is not used.